packages/icalendar/type/Recur.pkl

```diff
@@ -0,0 +1,198 @@
+/// If the "DTSTART" property is specified as a date with local time, then the UNTIL rule part MUST also be specified as a date with local time.
+/// If the "DTSTART" property is specified as a date with UTC time or a date with local time and time zone reference, then the UNTIL rule part MUST be specified as a date with UTC time.
+/// If not present, and the COUNT rule part is also not present, the "RRULE" is considered to repeat forever.
+until: (
+  * DateTime(
+      // if dtstart is defined and a datetime, until must also be a datetime if defined
+      dtstart?.ifNonNull((it) -> it is DateTime) ?? true,
+      dtstart?.ifNonNull((it) -> utc == (it as DateTime).utc) ?? true,
+      utc.implies(tzid == null)
+    )
+  |Date(
+    // if dtstart is defined and a date, until must also be a date if defined
+    dtstart?.ifNonNull((it) -> it is Date) ?? true,
+    tzid == null
+  )
+  )?
+
+/// The COUNT rule part defines the number of occurrences at which to range-bound the recurrence.
+///
+/// The "DTSTART" property value always counts as the first occurrence.
+count: Int(this > 0)?
+
+/// The BYSECOND rule part specifies one or more seconds within a minute.
+bysecond: Listing<UInt(isBetween(0, 60))>(
+    !isEmpty,
+    !(dtstart is Date)
+  )?
+
+/// The BYMINUTE rule part specifies one or more minutes within an hour.
+byminute: Listing<UInt(isBetween(0, 59))>(
+    !isEmpty,
+    !(dtstart is Date)
+  )?
+
+/// The BYHOUR rule part specifies one or more hours of the day.
+byhour: Listing<UInt(isBetween(0, 23))>(
+    !isEmpty,
+    !(dtstart is Date)
+  )?
+
+/// The BYDAY rule part specifies one or more days of the week.
+///
+/// Each BYDAY value can also be preceded by a positive (+n) or negative (-n) integer.
+/// If present, this indicates the nth occurrence of a specific day within the MONTHLY or YEARLY "RRULE".
+/// For example, within a MONTHLY rule, +1MO (or simply 1MO) represents the first Monday within the month, whereas -1MO represents the last Monday of the month.
+/// The numeric value in a BYDAY rule part with the FREQ rule part set to YEARLY corresponds to an offset within the month when the BYMONTH rule part is present, and corresponds to an offset within the year when the BYWEEKNO or BYMONTH rule parts are present.
+///
+/// If an integer modifier is not present, it means all days of this type within the specified frequency.
+/// For example, within a MONTHLY rule, MO represents all Mondays within the month.
+byday: Listing<ByWeekday>(
+    !isEmpty,
+    freq == "MONTHLY" || freq == "YEARLY",
+    !(freq == "YEARLY" && byweekno != null)
+  )?
+
+/// The BYMONTHDAY rule part specifies one or more days of the month.
+///
+/// Negative values represent days counting back from the end of a month.
+/// For example, -10 represents the tenth to the last day of the month.
+bymonthday: Listing<Int(relativeRange(31))>(
```

Comment:
For each of these, I think it's a little better to split these into separate constraints, because Pkl gives a better error message when this fails.

Otherwise, you get a message like ``Type constraint `relativeRange(31)` failed. Value: 0``, and this doesn't make too much sense to users.

```suggestion
bymonthday: Listing<Int(isBetween(-31, 31), this != 0)>(
```